---
title: "Multi-Omic Strategies for TWAS (MOSTWAS)"
author: "Arjun Bhattacharya"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction to transcriptome-wide association studies (TWAS)

TWAS are techniques that were concurrently
created by [Gamazon et al, 2015](https://www.ncbi.nlm.nih.gov/pubmed/26258848)
and [Gusev et al, 2016](https://www.ncbi.nlm.nih.gov/pubmed/26854917)
to leverage eQTLs to identify gene-trait associations.
A traditional TWAS pipeline is as follows:

1. In a tissue that is relevant to a phenotype of interest (i.e. breast tumor
for breast cancer outcomes or the pre-frontal cortex for neuropsychiatric disorders),
we train predictive models of tissue-specific expression from genotype.
2. Then, in a large GWAS cohort, we use these predictive models
to impute tissue-specific genetically regulated expression to detect
gene-trait associations.

In training predictive expression models, most TWAS methods (e.g. [PrediXcan](https://www.ncbi.nlm.nih.gov/pubmed/26258848), 
[FUSION](https://www.ncbi.nlm.nih.gov/pubmed/26854917), 
[TIGAR](https://www.ncbi.nlm.nih.gov/pubmed/31230719), 
[EpiXcan](https://www.nature.com/articles/s41467-019-11874-7), etc)
focus on local genetic variation around the gene of interest (0.5 to 1 Megabase
around the gene). However, [Boyle et al, 2017](https://www.ncbi.nlm.nih.gov/pubmed/28622505) and 
[Liu et al, 2019](https://www.ncbi.nlm.nih.gov/pubmed/31051098) have proposed 
that up to 70% of gene expression can be attributed
to distal variation in the genome, suggesting
that the inclusion of these distal variants
in TWAS models may improve prediction and power to detect
gene-trait associations.

For this reason, we have developed **M**ulti-**O**mic
**S**trategies for **TWAS** (**MOSTWAS**),
an intuitive suite of tools to prioritize distal
variants in transcriptomic prediction and conduct
TWAS-like association testing using GWAS summary statistics.
MOSTWAS incorporates two
methods to include distal-eQTLs
in transcriptomic prediction: Mediator-enriched
TWAS (MeTWAS) and distal-eQTL prioritization
via mediation analysis (DePMA). These
methods are summarized graphically
below.

```{r, out.width = "700px", echo=F}
knitr::include_graphics("mostwas_scheme.png")
```

* MeTWAS first
identifies $m$ mediators
(e.g. CpG sites,
miRNAs, gene coding
for transcription
factors, etc) such
that the intensity (methylation
or expression levels) 
of these mediators are associated
to the mRNA expression. A model
for the
genetically regulated intensities
(GRIn)
is estimated using
the local SNPs
to these mediators, and
the GRIn of these mediators
are imputed into the training
set. The final gene expression model
is estimated by incorporating
the GRIn of the mediators
as fixed effects and 
the local SNPs 
to the gene as regularized
effects (see Methods
for more details).

* DePMA
first identified
testing triplets
of the gene of
interest, a distal
eSNP (SNP
in an eQTL) to the gene,
and any associated mediators local
to the eQTL. We estimate
the total mediation effect (TME) 
of the eQTL on the gene
through the set of mediators
and test the two-sided
hypothesis of $H_0: TME = 0$.
Any distal-eSNP with a significant
TME is included with the SNPs
local to the gene of interest
to form the final design matrix.
A model including all
local SNPs and all significant
distal-eSNPs is fit to
the expression of the gene
using either elastic net
or linear mixed modeling (see Methods
for more details).

# Using MOSTWAS

MOSTWAS is meant to be run on a high-performance cluster (HPC)
and integrates with [PLINK](http://zzz.bwh.harvard.edu/plink/) 
and [GCTA](https://cnsgenomics.com/software/gcta/) to
perform file formatting, linkage disequilibrium estimation,
and heritability estimation. If your HPC already has
PLINK and GCTA installed, be sure to load and add those
modules to your personal PATH with something like this:

```{r, eval=F}
module add plink
module add gcta
```

If PLINK and GCTA aren't automatically installed,
you can simply download and install these software
add them manually to your PATH:

```{r, eval=F}
module use /path/to/personal/plink
module use /path/to/persona/gcta
module load plink
module load gcta64
module add plink
module add gcta
```

Now, we can go to `R` to use MOSTWAS.

## Installing MOSTWAS

MOSTWAS is available freely on Github.
To install the package:

```{r, out.width = "700px", eval=F}
devtools::install_github('bhattacharya-a-bt/MOSTWAS')
```

## Necessary input files

Let's assume that the reference eQTL panel has $n$ samples.
MOSTWAS, at the very least, requires the input
files for the following:

1. SNP dosages and locations,
2. gene expression and locations,
3. mediator intensities and locations, and
4. relevant covariates.

Mediators can include any biomarker that may
influence the transcription of a gene (i.e.
transcription factors, microRNAs, DNA CpG methylation sites,
etc). The term "intensity" here is used as a catch-all term
for expression and methylation of mediators.

MOSTWAS uses files that are formatted
in the `MatrixEQTL` format (see the
[MatrixEQTL](http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/) 
documentation). Briefly, SNP dosage, gene expression,
and mediator intensity files all have $n+1$ columns, where
the first column is an identifier for the biomarker (SNP, gene, or mediator)
and the last $n$ columns are the scaled values for that biomarker.
The location files have columns (in order) for the identifier,
the chromosome, and the base-pair position of the biomarker.
Gene location files have a fourth column with the end location of the gene.

We recommend that the file that contains the covariates include principal
components of the SNP dosage matrix (in the range of 5-20, depending
on the dataset). Relevant clinical covariates can be included as well.
This file has $n+1$ columns, where the first column is the identifier
for the covariate and the last $n$ columns contain
the values for the covariates across the $n$ samples.

## eQTL analysis


