<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-Omic Strategies for Transcriptome-Wide Association Studies • MOSTWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Multi-Omic Strategies for Transcriptome-Wide Association Studies">
<meta property="og:description" content="Enrichment for transcriptome imputation and association studies using mediating biomarkers.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MOSTWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/MOSTWAS_vignette.html">Multi-Omic Strategies for TWAS (MOSTWAS)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="multi-omic-strategies-for-transcriptome-wide-association-studies-mostwas" class="section level2">
<h2 class="hasAnchor">
<a href="#multi-omic-strategies-for-transcriptome-wide-association-studies-mostwas" class="anchor"></a>Multi-Omic Strategies for Transcriptome-Wide Association Studies (MOSTWAS)</h2>
<p>For a full vignette on using MOSTWAS, click <a href="http://htmlpreview.github.io/?https://github.com/bhattacharya-a-bt/MOSTWAS/blob/master/vignettes/MOSTWAS_vignette.html">here</a>.</p>
<p>TWAS are techniques that were concurrently created by <a href="https://www.ncbi.nlm.nih.gov/pubmed/26258848">Gamazon et al, 2015</a> and <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854917">Gusev et al, 2016</a> to leverage eQTLs to identify gene-trait associations. A traditional TWAS pipeline is as follows:</p>
<ol>
<li>In a tissue that is relevant to a phenotype of interest (i.e. breast tumor for breast cancer outcomes or the pre-frontal cortex for neuropsychiatric disorders), we train predictive models of tissue-specific expression from genotype.</li>
<li>Then, in a large GWAS cohort, we use these predictive models to impute tissue-specific genetically regulated expression to detect gene-trait associations.</li>
</ol>
<p>In training predictive expression models, most TWAS methods (e.g. <a href="https://www.ncbi.nlm.nih.gov/pubmed/26258848">PrediXcan</a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854917">FUSION</a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/31230719">TIGAR</a>, <a href="https://www.nature.com/articles/s41467-019-11874-7">EpiXcan</a>, etc) focus on local genetic variation around the gene of interest (0.5 to 1 Megabase around the gene). However, <a href="https://www.ncbi.nlm.nih.gov/pubmed/28622505">Boyle et al, 2017</a> and <a href="https://www.ncbi.nlm.nih.gov/pubmed/31051098">Liu et al, 2019</a> have proposed that up to 70% of gene expression can be attributed to distal variation in the genome, suggesting that the inclusion of these distal variants in TWAS models may improve prediction and power to detect gene-trait associations.</p>
<p>For this reason, we have developed <strong>M</strong>ulti-<strong>O</strong>mic <strong>S</strong>trategies for <strong>TWAS</strong> (<strong>MOSTWAS</strong>), an intuitive suite of tools to prioritize distal variants in transcriptomic prediction and conduct TWAS-like association testing using GWAS summary statistics. MOSTWAS incorporates two methods to include distal-eQTLs in transcriptomic prediction: Mediator-enriched TWAS (MeTWAS) and distal-eQTL prioritization via mediation analysis (DePMA). These methods are summarized graphically below.</p>
<p><img src="articles/mostwas_scheme.png"></p>
<ul>
<li><p>MeTWAS first identifies <em>m</em> mediators (e.g. CpG sites, miRNAs, gene coding for transcription factors, etc) such that the intensity (methylation or expression levels) of these mediators are associated to the mRNA expression. A model for the genetically regulated intensities (GRIn) is estimated using the local SNPs to these mediators, and the GRIn of these mediators are imputed into the training set. The final gene expression model is estimated by incorporating the GRIn of the mediators as fixed effects and the local SNPs to the gene as regularized effects (see Methods for more details).</p></li>
<li><p>DePMA first identified testing triplets of the gene of interest, a distal eSNP (SNP in an eQTL) to the gene, and any associated mediators local to the eQTL. We estimate the total mediation effect (TME) of the eQTL on the gene through the set of mediators and test the two-sided hypothesis that <em>TME = 0</em>. Any distal-eSNP with a significant TME is included with the SNPs local to the gene of interest to form the final design matrix. A model including all local SNPs and all significant distal-eSNPs is fit to the expression of the gene using either elastic net or linear mixed modeling (see Methods for more details).</p></li>
</ul>
<p>All MOSTWAS models output:</p>
<ul>
<li>the SNP-gene effect sizes (i.e. the model),</li>
<li>predictive performance in cross-validation,</li>
<li>heritability estimate from GCTA,</li>
<li>
<em>P</em>-value for the likelihood ratio test for heritability, and</li>
<li>predicted expression in the reference panel.</li>
</ul>
<p>We recommend training genes with both MeTWAS and DePMA and prioritizing for association testing the model with better cross-validation performance.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Arjun Bhattacharya <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Arjun Bhattacharya.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
