<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-Omic Strategies for TWAS (MOSTWAS) • MOSTWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-Omic Strategies for TWAS (MOSTWAS)">
<meta property="og:description" content="MOSTWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MOSTWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MOSTWAS_vignette.html">Multi-Omic Strategies for TWAS (MOSTWAS)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="MOSTWAS_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="MOSTWAS_vignette_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="MOSTWAS_vignette_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-Omic Strategies for TWAS (MOSTWAS)</h1>
            
      
      
      <div class="hidden name"><code>MOSTWAS_vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-transcriptome-wide-association-studies-twas" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-transcriptome-wide-association-studies-twas" class="anchor"></a>Introduction to transcriptome-wide association studies (TWAS)</h2>
<p>TWAS are techniques that were concurrently created by <a href="https://www.ncbi.nlm.nih.gov/pubmed/26258848">Gamazon et al, 2015</a> and <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854917">Gusev et al, 2016</a> to leverage eQTLs to identify gene-trait associations. A traditional TWAS pipeline is as follows:</p>
<ol style="list-style-type: decimal">
<li>In a tissue that is relevant to a phenotype of interest (i.e. breast tumor for breast cancer outcomes or the pre-frontal cortex for neuropsychiatric disorders), we train predictive models of tissue-specific expression from genotype.</li>
<li>Then, in a large GWAS cohort, we use these predictive models to impute tissue-specific genetically regulated expression to detect gene-trait associations.</li>
</ol>
<p>In training predictive expression models, most TWAS methods (e.g. <a href="https://www.ncbi.nlm.nih.gov/pubmed/26258848">PrediXcan</a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854917">FUSION</a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/31230719">TIGAR</a>, <a href="https://www.nature.com/articles/s41467-019-11874-7">EpiXcan</a>, etc) focus on local genetic variation around the gene of interest (0.5 to 1 Megabase around the gene). However, <a href="https://www.ncbi.nlm.nih.gov/pubmed/28622505">Boyle et al, 2017</a> and <a href="https://www.ncbi.nlm.nih.gov/pubmed/31051098">Liu et al, 2019</a> have proposed that up to 70% of gene expression can be attributed to distal variation in the genome, suggesting that the inclusion of these distal variants in TWAS models may improve prediction and power to detect gene-trait associations.</p>
<p>For this reason, we have developed <strong>M</strong>ulti-<strong>O</strong>mic <strong>S</strong>trategies for <strong>TWAS</strong> (<strong>MOSTWAS</strong>), an intuitive suite of tools to prioritize distal variants in transcriptomic prediction and conduct TWAS-like association testing using GWAS summary statistics. MOSTWAS incorporates two methods to include distal-eQTLs in transcriptomic prediction: Mediator-enriched TWAS (MeTWAS) and distal-eQTL prioritization via mediation analysis (DePMA).</p>
<ul>
<li><p>MeTWAS first identifies <span class="math inline">\(m\)</span> mediators (e.g. CpG sites, miRNAs, gene coding for transcription factors, etc) such that the intensity (methylation or expression levels) of these mediators are associated to the mRNA expression. A model for the genetically regulated intensities (GRIn) is estimated using the local SNPs to these mediators, and the GRIn of these mediators are imputed into the training set. The final gene expression model is estimated by incorporating the GRIn of the mediators as fixed effects and the local SNPs to the gene as regularized effects (see Methods for more details).</p></li>
<li><p>DePMA first identified testing triplets of the gene of interest, a distal eSNP (SNP in an eQTL) to the gene, and any associated mediators local to the eQTL. We estimate the total mediation effect (TME) of the eQTL on the gene through the set of mediators and test the two-sided hypothesis of <span class="math inline">\(H_0: TME = 0\)</span>. Any distal-eSNP with a significant TME is included with the SNPs local to the gene of interest to form the final design matrix. A model including all local SNPs and all significant distal-eSNPs is fit to the expression of the gene using either elastic net or linear mixed modeling (see Methods for more details).</p></li>
</ul>
<p>All MOSTWAS models output:</p>
<ul>
<li>the SNP-gene effect sizes (i.e. the model),</li>
<li>cross-validation <span class="math inline">\(R^2\)</span>,</li>
<li>heritability estimate from GCTA,</li>
<li>
<span class="math inline">\(P\)</span>-value for the likelihood ratio test for heritability, and</li>
<li>predicted expression in the reference panel.</li>
</ul>
<p>We recommend training genes with both MeTWAS and DePMA and prioritizing the model with greater cross-validation <span class="math inline">\(R^2\)</span> for association testing.</p>
</div>
<div id="using-mostwas" class="section level2">
<h2 class="hasAnchor">
<a href="#using-mostwas" class="anchor"></a>Using MOSTWAS</h2>
<p>MOSTWAS is meant to be run on a high-performance cluster (HPC) and integrates with <a href="https://cnsgenomics.com/software/gcta/">GCTA</a> to perform file formatting, linkage disequilibrium estimation, and heritability estimation. If your HPC already has PLINK and GCTA installed, be sure to load and add those modules to your personal PATH with something like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>module add plink</span>
<span id="cb1-2"><a href="#cb1-2"></a>module add gcta</span></code></pre></div>
<p>If PLINK and GCTA aren’t automatically installed, you can simply download and install these software add them manually to your PATH:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>module use <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>personal<span class="op">/</span>plink</span>
<span id="cb2-2"><a href="#cb2-2"></a>module use <span class="op">/</span>path<span class="op">/</span>to<span class="op">/</span>persona<span class="op">/</span>gcta</span>
<span id="cb2-3"><a href="#cb2-3"></a>module load plink</span>
<span id="cb2-4"><a href="#cb2-4"></a>module load gcta64</span>
<span id="cb2-5"><a href="#cb2-5"></a>module add plink</span>
<span id="cb2-6"><a href="#cb2-6"></a>module add gcta</span></code></pre></div>
<p>Now, we can go to <code>R</code> to use MOSTWAS.</p>
<div id="installing-mostwas" class="section level3">
<h3 class="hasAnchor">
<a href="#installing-mostwas" class="anchor"></a>Installing MOSTWAS</h3>
<p>MOSTWAS is available freely on Github. To install the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'bhattacharya-a-bt/MOSTWAS'</span><span class="op">)</span></code></pre></div>
</div>
<div id="necessary-input-files" class="section level3">
<h3 class="hasAnchor">
<a href="#necessary-input-files" class="anchor"></a>Necessary input files</h3>
<p>Let’s assume that the reference eQTL panel has <span class="math inline">\(n\)</span> samples. MOSTWAS, at the very least, requires the input files for the following:</p>
<ol style="list-style-type: decimal">
<li>SNP dosages and locations,</li>
<li>gene expression and locations,</li>
<li>mediator intensities and locations, and</li>
<li>relevant covariates.</li>
</ol>
<p>Mediators can include any biomarker that may influence the transcription of a gene (i.e. transcription factors, microRNAs, DNA CpG methylation sites, etc). The term “intensity” here is used as a catch-all term for expression and methylation of mediators.</p>
<p>MOSTWAS uses files that are formatted in the <code>MatrixEQTL</code> format (see the <a href="http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/">MatrixEQTL</a> documentation). Briefly, SNP dosage, gene expression, and mediator intensity files all have <span class="math inline">\(n+1\)</span> columns, where the first column is an identifier for the biomarker (SNP, gene, or mediator) and the last <span class="math inline">\(n\)</span> columns are the scaled values for that biomarker. The location files have columns (in order) for the identifier, the chromosome, and the base-pair position of the biomarker. Gene location files have a fourth column with the end location of the gene.</p>
<p>We recommend that the file that contains the covariates include principal components of the SNP dosage matrix (in the range of 5-20, depending on the dataset). Relevant clinical covariates can be included as well. This file has <span class="math inline">\(n+1\)</span> columns, where the first column is the identifier for the covariate and the last <span class="math inline">\(n\)</span> columns contain the values for the covariates across the <span class="math inline">\(n\)</span> samples.</p>
<p>Here’s an example of what a SNP dosage file looks like:</p>
<pre><code>#&gt;       SNP     sample1      sample2    sample3    sample4    sample5     sample6
#&gt; 1:  snp55  0.71266631 -0.008309014 -1.1007406  0.1860351 -1.2589214  0.24350206
#&gt; 2: snp148 -0.07356440  0.128855402  0.7145094  0.1764178 -0.2153845  1.11436266
#&gt; 3: snp211 -0.03763417 -0.145875628 -0.2464703  0.9158482 -2.4719617 -0.07655271
#&gt; 4:  snp79 -0.68166048 -0.163910957 -0.3197862  0.3201767 -0.6741693  0.93292505
#&gt; 5:  snp61 -0.32427027  1.763552003  1.3626443 -0.3666873 -0.5012972 -0.15798174
#&gt; 6:  snp56  0.06016044  0.762586512 -1.2278826 -0.9406113  1.5423258 -0.68150951
#&gt;       sample7    sample8     sample9
#&gt; 1: -1.5894114 -0.5572063  0.04947015
#&gt; 2:  1.2341042 -1.1133253 -0.45554882
#&gt; 3: -0.3354548 -1.6306354  0.65788940
#&gt; 4: -2.0010025 -0.3878239 -1.48546741
#&gt; 5: -0.0174584  1.7226320  1.46918455
#&gt; 6: -0.2409669 -0.5585003 -0.22176117</code></pre>
<p>Here’s an example of what a SNP location file looks like:</p>
<pre><code>#&gt;    snpid chr        pos
#&gt; 1:  snp1  10 8066096444
#&gt; 2:  snp2  10 8066309675
#&gt; 3:  snp3  10 8066711134
#&gt; 4:  snp4  10 8067381843
#&gt; 5:  snp5  10 8065968791
#&gt; 6:  snp6  10 8067362206</code></pre>
<p>For training models using the <code><a href="../reference/MeTWAS.html">MeTWAS()</a></code> and <code><a href="../reference/DePMA.html">DePMA()</a></code> models, we recommend keeping the genotype files in binary PLINK format (i.e. .bed/.bim/.fam) and reading in the file using the <code>bigsnpr</code> package. Here’s a link to their <a href="https://privefl.github.io/bigsnpr/">documentation</a>.</p>
</div>
<div id="eqtl-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#eqtl-analysis" class="anchor"></a>eQTL analysis</h3>
<p>MeTWAS and DePMA require results for QTL analysis:</p>
<ul>
<li>MeTWAS requires associations between mediators and genes, and</li>
<li>DePMA requires distal-eQTLs between SNPs and genes, and local mediator QTLs for those distal-eQTLs.</li>
</ul>
<p>We have provided a wrapper function to do these eQTL analyses with <code>MatrixEQTL</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/eQTL_MOSTWAS.html">eQTL_MOSTWAS</a></span><span class="op">(</span>SNP_file_name <span class="op">=</span> <span class="st">'snps.txt'</span>,
             snps_location_file_name <span class="op">=</span> <span class="st">'snpLocs.txt'</span>,
             SNP_file_name_dis <span class="op">=</span> <span class="st">'snps_dis.txt'</span>,
             expression_file_name <span class="op">=</span> <span class="st">'genes.txt'</span>,
             gene_location_file_name <span class="op">=</span> <span class="st">'geneLocs.txt'</span>,
             mediators_file_name <span class="op">=</span> <span class="st">'mediators.txt'</span>,
             meds_location_file_name <span class="op">=</span> <span class="st">'medLocs.txt'</span>,
             covariates_file_name <span class="op">=</span> <span class="st">'covs.txt'</span>,
             output_file_name_loc_qtl <span class="op">=</span> <span class="st">'out_loc_qtl.txt'</span>,
             output_file_name_dis_qtl <span class="op">=</span> <span class="st">'out_dis_qtl.txt'</span>,
             output_file_name_loc_med <span class="op">=</span> <span class="st">'out_loc_med.txt'</span>,
             output_file_name_dis_med <span class="op">=</span> <span class="st">'out_dos_med.txt'</span>,
             p_loc_qtl <span class="op">=</span> <span class="fl">1e-6</span>,
             p_dis_qtl <span class="op">=</span> <span class="fl">1e-6</span>,
             FDRcut <span class="op">=</span> <span class="fl">0.05</span>,
             useModel <span class="op">=</span> <span class="st">'modelLinear'</span>,
             DePMA <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>We have provided a toggle <code>DePMA</code> that takes in a logical object (i.e. <code>TRUE</code> or <code>FALSE</code>) to do either QTL analysis for DePMA or MeTWAS. <code>MatrixEQTL</code> is also a very intuitive software, so you can easily code your own analyses.</p>
<p><strong>Please cite <a href="https://www.ncbi.nlm.nih.gov/pubmed/22492648">Shabalin 2012</a> if you use MOSTWAS!</strong></p>
</div>
<div id="training-models-with-metwas" class="section level3">
<h3 class="hasAnchor">
<a href="#training-models-with-metwas" class="anchor"></a>Training models with MeTWAS</h3>
<p>MOSTWAS wrapper functions have multiple options to provide users with freedom in training predictive models. It may get a little confusing, but we’ll cover the options one-by-one.</p>
<p>You can run MeTWAS all in one, for a given gene (call this gene <code>test</code>):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/MeTWAS.html">MeTWAS</a></span><span class="op">(</span><span class="va">geneInt</span>,
       <span class="va">snpObj</span>,
       <span class="va">mediator</span>,
       <span class="va">medLocs</span>,
       <span class="va">covariates</span>,
       <span class="va">dimNumeric</span>,
       <span class="va">qtlFull</span>,
       h2Pcutoff <span class="op">=</span> <span class="fl">.1</span>,
       numMed <span class="op">=</span> <span class="fl">5</span>,
       seed <span class="op">=</span> <span class="fl">1218</span>,
       k <span class="op">=</span> <span class="fl">5</span>,
       cisDist <span class="op">=</span> <span class="fl">1e6</span>,
       parallel <span class="op">=</span> <span class="cn">F</span>,
       prune <span class="op">=</span> <span class="cn">F</span>,
       ldThresh <span class="op">=</span> <span class="fl">.5</span>,
       <span class="va">cores</span>,
       verbose <span class="op">=</span> <span class="cn">T</span>,
       R2Cutoff <span class="op">=</span> <span class="fl">0.01</span>,
       <span class="va">modelDir</span>,
       <span class="va">tempFolder</span><span class="op">)</span></code></pre></div>
<p>A few notes on these options:</p>
<ul>
<li>
<code>snpObj</code> is a bigSNP object that is read in using the <code>bigsnpr</code> package. This is a much more memory-efficient and faster approach.</li>
<li>
<code>dimNumeric</code> takes in the number of leading rows of the <code>covariates</code> file that are genotype PCs for heritability estimation using GCTA.</li>
<li>
<code>qtlFull</code> takes in the QTL results, as a <code>data.frame</code> or <code>data.table</code> that give associations between mediators and genes. This is in the format of MatrixEQTL output.</li>
<li>
<code>h2Pcutoff</code> gives a <span class="math inline">\(P\)</span>-value cutoff for heritability estimation using GCTA</li>
<li>
<code>numMed</code> gives an upper limit to the number of top gene-associated mediators to limit computational time. Ideally, in the distal QTL analysis between mediators (predictors) and genes (outcomes)</li>
<li>
<code>parallel</code> toggles parallel fitting of the <code>numMed</code> local-only models of mediators with an <code>mclapply()</code> implementation. The <code>cores</code> option allows you to set the number of parallel cores for this implementation.</li>
<li>
<code>prune</code> togggles wheter the SNP dosages need to be LD-pruned. We have seen previously that LD-pruning may improve prediction <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1942-6">Bhattacharya et al 2020</a>. <code>ldThresh</code> provides the LD threshold for clumping.</li>
<li>
<code>snpAnnot</code> takes in a data.frame of SNP annotations that includes three columns. This is generally needed if the SNP identifiers are not in 1000GP form (i.e. chrom:position:REF:ALT).</li>
<li>Models are outputted to <code>modelDir</code> and backing files for <code>bigsnpr</code> packages are stored in <code>tempFolder</code>
</li>
</ul>
</div>
<div id="training-models-with-depma" class="section level3">
<h3 class="hasAnchor">
<a href="#training-models-with-depma" class="anchor"></a>Training models with DePMA</h3>
<p>You can run DePMA all in one, for a given gene (call this gene <code>test</code>). Many of the options are similar, so we’ll highlight the different options in DePMA:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/DePMA.html">DePMA</a></span><span class="op">(</span>geneInt <span class="op">=</span> <span class="st">'test'</span>,
      <span class="va">snpObj</span>,
      <span class="va">mediator</span>,
      <span class="va">medLocs</span>,
      <span class="va">covariates</span>,
      cisDist <span class="op">=</span> <span class="fl">1e6</span>,
      <span class="va">qtlTra</span>,
      <span class="va">qtMed</span>,
      <span class="va">h2Pcutoff</span>,
      <span class="va">dimNumeric</span>,
      <span class="va">verbose</span>,
      <span class="va">seed</span>,
      sobel <span class="op">=</span> <span class="cn">F</span>,
      nperms <span class="op">=</span> <span class="fl">1000</span>,
      <span class="va">k</span>,
      <span class="va">parallel</span>,
      parType <span class="op">=</span> <span class="st">'no'</span>,
      <span class="va">prune</span>,
      ldThresh <span class="op">=</span> <span class="fl">.5</span>,
      <span class="va">cores</span>,
      <span class="va">qtlTra_parts</span>,
      <span class="va">qtMed_parts</span>,
      <span class="va">modelDir</span>,
      <span class="va">tempFolder</span>,
      <span class="va">R2Cutoff</span><span class="op">)</span></code></pre></div>
<p>A few notes on these options:</p>
<ul>
<li>
<code>qtlTra</code> and <code>qtMed</code> take in outputs from <code>eQTL_MOSTWAS</code> if the <code>DePMA = T</code> toggle is set. These are <code>data.frame</code> or <code>data.table</code> objects that contain the MatrixEQTL output for associations between distal SNPs and genes and those distal-eQTLs and any local mediators. For cross-validation (generally <span class="math inline">\(k &lt; 5\)</span> folds), you must also run fold-wise QTL analysis and input these filenames in <code>qtlTra_parts</code> and <code>qtMed_parts</code> as character vectors.</li>
<li>
<code>sobel</code> toggles whether the asymptotic Sobel test is used, instead of the permutation test, to test if the total mediation effect for a distal-eQTL is large. The asymptotic Sobel test is less powerful yet much faster compared to the permutation test.</li>
<li>We use the <code>boot</code> package for permutation. <code>nperms</code> sets the number of resamplings in permutation test, and <code>parType</code> sets the parallelization method as in the <code>boot</code> package.</li>
</ul>
</div>
<div id="running-multiple-genes-in-parallel" class="section level3">
<h3 class="hasAnchor">
<a href="#running-multiple-genes-in-parallel" class="anchor"></a>Running multiple genes in parallel</h3>
<p>Usually, you’ll need to train models for 15,000+ genes in an RNA-seq panel. Serially running DePMA is slow (MeTWAS is quite fast), even when parallelizing within a single gene. We recommend a batch submission procedure, perhaps using the <code>rslurm</code> package. Given that the genes of interest as in the vector <code>geneList</code>, we can simply define a <code>data.frame</code> of parameters and submit multiple jobs all in one:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pars</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>geneInt <span class="op">=</span> <span class="va">geneList</span><span class="op">)</span>
<span class="va">batch_DeP</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">geneInt</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="../reference/DePMA.html">DePMA</a></span><span class="op">(</span><span class="va">geneInt</span>,<span class="va">snpObj</span>,<span class="va">mediator</span>,<span class="va">...</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rslurm</span><span class="op">)</span>
<span class="va">sjob</span> <span class="op">&lt;-</span> <span class="fu">slurm_apply</span><span class="op">(</span><span class="va">batch_DeP</span>, <span class="va">pars</span>, jobname <span class="op">=</span> <span class="st">'test_DePMA'</span>,
                    nodes <span class="op">=</span> <span class="fl">2</span>, cpus_per_node <span class="op">=</span> <span class="fl">2</span>, submit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>This generates <code>R</code> scripts for every gene and submits them to your SLURM cluster. You can delete all the files using <code>cleanup_files(sjob)</code>.</p>
<p>Other options include the <code>BatchJobs</code> <a href="https://cran.r-project.org/web/packages/BatchJobs/BatchJobs.pdf">package on CRAN</a> or <a href="https://snakemake.readthedocs.io/en/stable/">snakemake</a>.</p>
<p>With <code>slurm_apply()</code>, we were able to fit a whole RNA-seq panel using both MeTWAS and DePMA in roughly 20 hours.</p>
</div>
<div id="fusion-like-association-testing" class="section level3">
<h3 class="hasAnchor">
<a href="#fusion-like-association-testing" class="anchor"></a>FUSION-like association testing</h3>
<p>We provide functions for testing as well. The function <code><a href="../reference/burdenTest.html">burdenTest()</a></code> does the weighted burned test of association and gives the option for permutation testing, as in <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854917">FUSION</a>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">burdenTest</span>(wgt, <span class="co">### .RData file name for a given gene's MOSTWAS model</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>           snps, <span class="co">### data.frame/data.table for SNP dosages</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>           sumStats, <span class="co">### data.frame/data.table for GWAS summary statistics</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>           <span class="dt">snpAnnot =</span> <span class="ot">NULL</span>,</span>
<span id="cb10-5"><a href="#cb10-5"></a>           <span class="dt">onlyCis =</span> F, <span class="co">### toggle to include only local SNPs</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>           beta, <span class="co">### column name for effect sizes in GWAS summary stats</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>           se, <span class="co">### column name for standard errors for betas in GWAS summary stats</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>           chr, <span class="co">### column name for SNP chromosome locations in GWAS summary stats</span></span>
<span id="cb10-9"><a href="#cb10-9"></a>           pos, <span class="co">### column name for SNP basepair position in GWAS summary stats</span></span>
<span id="cb10-10"><a href="#cb10-10"></a>           ref, <span class="co">### column name for SNP reference allele in GWAS summary stats</span></span>
<span id="cb10-11"><a href="#cb10-11"></a>           pval, <span class="co">### column name for P-value in GWAS summary stats</span></span>
<span id="cb10-12"><a href="#cb10-12"></a>           <span class="dt">R2cutoff =</span> <span class="fl">0.01</span>, <span class="co">### cross-validation R2 cutoff to test a gene</span></span>
<span id="cb10-13"><a href="#cb10-13"></a>           alpha, <span class="co">### P-value cutoff in TWAS association to conduct permuation test</span></span>
<span id="cb10-14"><a href="#cb10-14"></a>           <span class="dt">nperms =</span> <span class="fl">1e3</span> <span class="co">### number of permutations in permuation test)</span></span></code></pre></div>
<p>A comment on permutation testing:</p>
<ul>
<li>Permutation testing assess wheter the TWAS association persists when conditioning on large GWAS effect sizes. This test is meant to be conservatively used, so we recommend imposing a stringent <span class="math inline">\(P\)</span>-value cutoff to prioritize a gene for permutation testing (i.e. <code>alpha</code> at the Bonferroni cutoff).</li>
</ul>
</div>
<div id="distal-snps-added-last-test" class="section level3">
<h3 class="hasAnchor">
<a href="#distal-snps-added-last-test" class="anchor"></a>Distal-SNPs added last test</h3>
<p>For genes with significant TWAS associations from <code><a href="../reference/burdenTest.html">burdenTest()</a></code>, we may be interested in assessing whether the distal SNPs provide any added information beyond what we can observe from the local variation around the gene of interest. We developed an added-last test of distal SNPs using summary statistics. It can be run using the <code><a href="../reference/addedLastTest.html">addedLastTest()</a></code> function that take in similar options as <code><a href="../reference/burdenTest.html">burdenTest()</a></code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">addedLastTest</span>(wgt,</span>
<span id="cb11-2"><a href="#cb11-2"></a>              snps,</span>
<span id="cb11-3"><a href="#cb11-3"></a>              sumStats,</span>
<span id="cb11-4"><a href="#cb11-4"></a>              <span class="dt">snpAnnot =</span> <span class="ot">NULL</span>,</span>
<span id="cb11-5"><a href="#cb11-5"></a>              beta,</span>
<span id="cb11-6"><a href="#cb11-6"></a>              se,</span>
<span id="cb11-7"><a href="#cb11-7"></a>              chr,</span>
<span id="cb11-8"><a href="#cb11-8"></a>              pos,</span>
<span id="cb11-9"><a href="#cb11-9"></a>              ref,</span>
<span id="cb11-10"><a href="#cb11-10"></a>              pval,</span>
<span id="cb11-11"><a href="#cb11-11"></a>              R2cutoff,</span>
<span id="cb11-12"><a href="#cb11-12"></a>              locChrom <span class="co">### chromosome on which the gene of interest exists)</span></span></code></pre></div>
<p>The added-last test need not be run only on genes prioritized for permutation testing. We recommned any significant gene in overall TWAS association to be prioritized for added-last testing. The results from this test can allow users to focus on these distal regions for functional hypothesis generation and possible follow-up studies.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Arjun Bhattacharya.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
